table/fib-strategy-hashtable.go

```diff
@@ -0,0 +1,410 @@
+/* YaNFD - Yet another NDN Forwarding Daemon
+ *
+ * Copyright (C) 2022 Danning Yu.
+ *
+ * This file is licensed under the terms of the MIT License, as found in LICENSE.md.
+ */
+
+// This file implements a hash table version of the FIB, as outlined in section 2.3 of:
+// W. So, A. Narayanan and D. Oran,
+// "Named data networking on a router: Fast and DoS-resistant forwarding with hash tables,"
+// Architectures for Networking and Communications Systems, 2013, pp. 215-225,
+// doi: 10.1109/ANCS.2013.6665203.
+
+package table
+
+import (
+	"sync"
+
+	"github.com/cespare/xxhash"
+	"github.com/named-data/YaNFD/ndn"
+)
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+
+	return b
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+
+	return b
+}
+
+type virtualDetails struct {
+	// isVirtual marks whether this a virtual node, as defined in the paper.
+	// If there exists a name with length greater than M, then the node is
+	// marked as virtual.
+	isVirtual bool
```

Comment:
This is not used at all, since you separate the virtual table and the real table.

Reply:
No. I don't think the original paper separate virtual and real entries in different tables. Therefore, if an entry with length of M exists in table, that entry will be matched as both virtual and real entries.

Reply:
Hmm I notice that in my code I only ever set `isVirtual` but I never access it. Will remove.